src/kv/memstore.rs

```diff
@@ -24,47 +24,49 @@ impl Default for MemStore {
 #[async_trait(?Send)]
 impl Store for MemStore {
     async fn read<'a>(&'a self) -> Result<Box<dyn Read + 'a>> {
-        Ok(Box::new(ReadTransaction::new(self)))
+        let guard = self.map.read().await;
+        Ok(Box::new(ReadTransaction::new(guard)))
     }
 
     async fn write<'a>(&'a self) -> Result<Box<dyn Write + 'a>> {
-        Ok(Box::new(WriteTransaction::new(self)))
+        let guard = self.map.write().await;
+        Ok(Box::new(WriteTransaction::new(guard)))
     }
 }
 
 struct ReadTransaction<'a> {
-    store: &'a MemStore,
+    guard: RwLockReadGuard<'a, HashMap<String, Vec<u8>>>,
 }
 
 impl ReadTransaction<'_> {
-    fn new(store: &MemStore) -> ReadTransaction {
-        ReadTransaction { store }
+    fn new(guard: RwLockReadGuard<'_, HashMap<String, Vec<u8>>>) -> ReadTransaction {
+        ReadTransaction { guard }
     }
 }
 
 #[async_trait(?Send)]
 impl Read for ReadTransaction<'_> {
     async fn has(&self, key: &str) -> Result<bool> {
-        Ok(self.store.map.lock().await.contains_key(key))
+        Ok(self.guard.contains_key(key))
     }
 
     async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
-        match self.store.map.lock().await.get(key) {
+        match self.guard.get(key) {
             None => Ok(None),
             Some(v) => Ok(Some(v.to_vec())),
         }
     }
 }
 
 struct WriteTransaction<'a> {
-    rt: ReadTransaction<'a>,
+    guard: RwLockWriteGuard<'a, HashMap<String, Vec<u8>>>,
     pending: Mutex<HashMap<String, Option<Vec<u8>>>>,
 }
 
 impl WriteTransaction<'_> {
-    fn new(store: &MemStore) -> WriteTransaction {
+    fn new(guard: RwLockWriteGuard<'_, HashMap<String, Vec<u8>>>) -> WriteTransaction {
         WriteTransaction {
-            rt: ReadTransaction { store },
+            guard: guard,
```

Comment:
nit: you can just say `guard`. Also I think `cargo clippy` will whine about this.